### PR TITLE
Fix for Iterate

### PR DESCRIPTION
**Pull request by sppalkia**
This patch fixes a bug where `Iterate` expressions with nested `For` expressions could exhaust the stack by iterating via recursion. Specifically, the issue occurs if `For` loops are small enough that the runtime is not invoked and the loop body/continuation is called as function call.

This patch circumvents the problem by adding an "parallelize" annotation along with a transformation which forces calling into the parallel runtime for all for loops within an iterate. This could harm performance, so it needs a long term fix.

### CODE REVIEW

File: weld/sir.rs [hunk 1]
```diff
@@ -654,15 +656,15 @@ fn sir_param_correction(prog: &mut SirProgram) -> WeldResult<()> {
 }
 
 /// Convert an AST to a SIR program. Symbols must be unique in expr.
-pub fn ast_to_sir(expr: &TypedExpr) -> WeldResult<SirProgram> {
+pub fn ast_to_sir(expr: &TypedExpr, multithreaded: bool) -> WeldResult<SirProgram> {
```
Comment (sppalkia) [inline weld/sir.rs]: Added this to make it easier to do single vs. multithreaded optimizations in the SIR later.

**Comment by sppalkia** [comment]
@hvanhovell this should fix the seg fault in the wide aggregation workload. Addresses #244

File: weld/llvm.rs [hunk 2]
```diff
@@ -526,8 +533,17 @@ impl LlvmGenerator {
...
         let mut grain_size = DEFAULT_GRAIN_SIZE;
         if par_for.innermost {
-            ctx.code.add(format!("{} = icmp ule i64 {}, {}", bound_cmp, num_iters_str, grain_size));
-            ctx.code.add(format!("br i1 {}, label %for.ser, label %for.par", bound_cmp));
+            // Determine whether to always call parallel, always call serial, or
+            // choose based on the loop's size.
+            if par_for.parallelize {
+                ctx.code.add(format!("br label %for.par"));
+            }
+            else if self.multithreaded {
```
Comment (jjthomas) [inline weld/llvm.rs]: move this up a line so that it's on the same line as the previous closing curly brace

Comment (jjthomas) [reply weld/llvm.rs]: to make this code clearer i would actually have
`if par_for.parallelize {
...
} else {
  if (self.multithreaded) {
  ...
  } else {
  ...
  }
}`

File: weld/ast.rs [hunk 3]
```diff
@@ -276,6 +278,14 @@ impl Annotations {
         self.vectorize = Some(vectorize)
     }
 
+    pub fn parallelize(&self) -> &Option<bool> {
```
Comment (jjthomas) [inline weld/ast.rs]: i think `parallelize` is a misleading for this ... it should be something like `always_call_runtime` or `always_use_runtime`

**Comment by sppalkia** [comment]
Good point — do you think multithreaded should also be its own config?
Right now it just checks the number of threads set, but that’s slightly
different, since this is specifying how to compile the module (not how to
run it)

**Comment by jjthomas** [comment]
yeah setting the thread count and compile time and runtime is very confusing ... i think we should split up the conf into a CompileConf and RunConf, and thread count should only be in the RunConf. CompileConf can have something like a `compile serial` flag, which would emit just serial code, and then you could have a check at the start of the execution to make sure that the number of threads is always 1.

**Comment by sppalkia** [comment]
@jjthomas addressed your comments - yeah, might make more sense to have to Confs. For now I just made a separate option, I can break them up into compile/run in a different PR. We should also give the configuration keys consistent names.

**Comment by jjthomas** [comment]
looks good, feel free to merge

**Pull request merged**